File: src/runtime/java/mcp-mesh-sdk/src/main/java/io/mcpmesh/a2a/A2ABearer.java
package io.mcpmesh.a2a;

/**
 * Bearer-token credential for an outbound A2A call.
 *
 * <p>Provide either a literal token via {@link #of(String)} or the name
 * of an environment variable via {@link #fromEnv(String)}. The header is
 * resolved at call time (per request, just before the HTTP send) so
 * rotating the env var between calls picks up the new value without
 * reconstructing the surrounding {@link A2AClient}.
 *
 * <p>Mirrors {@code mesh._a2a_consumer.A2ABearer} from the Python
 * runtime (issue #908 Phase 1).
 *
 * <h2>Example</h2>
 * <pre>{@code
 * A2AClient client = new A2AClient(
 *     "http://upstream.example.com/agents/date",
 *     "get-date",
 *     A2ABearer.fromEnv("UPSTREAM_TOKEN"));
 * }</pre>
 */
public final class A2ABearer {

    private final String token;
    private final String tokenEnv;

    private A2ABearer(String token, String tokenEnv) {
        this.token = token;
        this.tokenEnv = tokenEnv;
    }

    /**
     * Use a literal bearer token.
     *
     * @param token the bearer credential, must be non-null and non-empty
     */
    public static A2ABearer of(String token) {
        if (token == null || token.isEmpty()) {
            throw new IllegalArgumentException("A2ABearer.of: token must be non-empty");
        }
        return new A2ABearer(token, null);
    }

    /**
     * Resolve the bearer token from the named environment variable on
     * each call.
     *
     * @param envVar the environment variable name, must be non-null and non-empty
     */
    public static A2ABearer fromEnv(String envVar) {
        if (envVar == null || envVar.isEmpty()) {
            throw new IllegalArgumentException("A2ABearer.fromEnv: envVar must be non-empty");
        }
        return new A2ABearer(null, envVar);
    }

    /**
     * Build the {@code Authorization} header value at call time.
     *
     * @return the {@code Bearer <token>} string suitable for
     *         {@code Authorization} headers.
     * @throws A2AAuthException if neither the literal token nor the
     *         configured env var resolves to a non-empty value.
     */
    public String authorizationHeader() {
        String resolved = token;
        if (resolved == null && tokenEnv != null) {
            resolved = System.getenv(tokenEnv);
        }
        if (resolved == null || resolved.isEmpty()) {
            throw new A2AAuthException(
                "A2ABearer: no token available (token_env="
                    + (tokenEnv == null ? "<unset>" : "'" + tokenEnv + "'")
                    + ", explicit_token="
                    + (token != null ? "set" : "unset")
                    + ")");
        }
        return "Bearer " + resolved;
    }
}

File: src/runtime/typescript/src/a2a/a2a-bearer.ts
/**
 * Bearer-token credential for an outbound A2A call (issue #917).
 *
 * Provide either `token` (literal) or `tokenEnv` (the name of an
 * environment variable). Resolution happens at header-build time so
 * rotating the env var between calls picks up the new value without
 * reconstructing the surrounding A2AClient. Mirrors
 * `mesh._a2a_consumer.A2ABearer` (Python) and `io.mcpmesh.a2a.A2ABearer`
 * (Java).
 */
import { A2AAuthError } from "./errors.js";

export interface A2ABearerConfig {
  /** Literal bearer token. Mutually exclusive with `tokenEnv`. */
  token?: string;
  /** Env-var name to read the token from on each call. Mutually exclusive with `token`. */
  tokenEnv?: string;
}

export class A2ABearer {
  private readonly token?: string;
  private readonly tokenEnv?: string;

  constructor(config: A2ABearerConfig) {
    if (!config.token && !config.tokenEnv) {
      throw new A2AAuthError(
        "A2ABearer: must specify either 'token' or 'tokenEnv'",
      );
    }
    if (config.token && config.tokenEnv) {
      throw new A2AAuthError(
        "A2ABearer: 'token' and 'tokenEnv' are mutually exclusive",
      );
    }
    if (config.token !== undefined && config.token.trim() === "") {
      throw new A2AAuthError("A2ABearer: 'token' must be non-blank");
    }
    if (config.tokenEnv !== undefined && config.tokenEnv.trim() === "") {
      throw new A2AAuthError("A2ABearer: 'tokenEnv' must be non-blank");
    }
    this.token = config.token;
    this.tokenEnv = config.tokenEnv;
  }

  /**
   * Build the value for the `Authorization` header at call time.
   *
   * Reads the env var (when configured) on every call so a rotated
   * credential is honoured mid-process.
   */
  authorizationHeader(): string {
    let resolved = this.token;
    if (resolved === undefined && this.tokenEnv) {
      resolved = process.env[this.tokenEnv];
    }
    if (!resolved || resolved.trim() === "") {
      throw new A2AAuthError(
        `A2ABearer: no token resolved (token=${this.token ? "set" : "unset"}, ` +
          `tokenEnv=${this.tokenEnv ?? "<unset>"})`,
      );
    }
    return `Bearer ${resolved}`;
  }
}

/**
 * Helper: normalise the user-friendly union (`A2ABearer | A2ABearerConfig`)
 * into a concrete `A2ABearer` instance, or `undefined` for "no auth".
 *
 * Lets the consumer-side `addTool({ a2aConfig: { auth: { tokenEnv: "X" } } })`
 * shortcut work without forcing the user to `new A2ABearer(...)` every time.
 */
export function resolveBearer(
  auth: A2ABearer | A2ABearerConfig | undefined,
): A2ABearer | undefined {
  if (auth === undefined) return undefined;
  if (auth instanceof A2ABearer) return auth;
  return new A2ABearer(auth);
}
